### DISCUSSION TITLE
Bundling exposes server side code in the client bundle?

### DISCUSSION THREAD

**Discussion by larsqa** [vercel/next.js#40763] [Help] [answered]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.6.0: Wed Aug 10 14:28:23 PDT 2022; root:xnu-8020.141.5~2/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.15.0
      npm: 8.5.5
      Yarn: 1.22.15
      pnpm: N/A
    Relevant packages:
      next: 12.2.2
      eslint-config-next: 12.2.5
      react: 18.0.0
      react-dom: 18.0.0

### My setup:
- I have a `/api.js` file, that contains api interfaces for both frontend (`clientApi`) and server (`serverApi`).
    - `clientApi` is used to make requests from the frontend to Next.js´s Api backend
    - `serverApi` is used to make in Next.js´s Api backend request to a third party api. The content of `serverApi` should never be exposed client side.

### The issue:
- importing `clientApi` in a `/pages/index.jsx` frontend page, only sometimes strips out the `serverApi` from the final bundle
-  Please see the attached github repo

### Expected Behavior

Next.js should be able to exclude `serverApi` from the final client side bundle in any case (unless it's seen as having side effects).

### Link to reproduction

https://github.com/larsqa/nextjs-internal-tree-shaking-example

### To Reproduce

Read readme of the linked github repo

**Comment by SukkaW** [comment, marked as answer]
The reason why the `serverApi` is included in the client bundle is that Webpack is not able to identify whether the `serverApi` or the `clientApi` is a pure export that doesn't have any side-effect.

The proper way to handle this is manually marking `serverApi` and `clientApi` as side-effect-free exports using `/*#__PURE__*/`, or splitting them into different files.

See also: https://webpack.js.org/guides/tree-shaking/#root

**Reply by larsqa** [reply]
I've read in other issues/discussions that adding `"sideEffects":false` to your `package.json` would solve the issue. (Ref: [#12557](https://github.com/vercel/next.js/issues/12557)). 

Since that didn't work, I didn't bother to test out `/*#__PURE__*/`, which surprisingly works! 
Any idea why?

For others reading here, declare your exports like
```jsx
export const serverApi = /*#__PURE__*/ createApi("I should only be available server side.")
```
